File: runtime/appruntime/metrics/prometheus/prometheus.go
package prometheus

import (
	"bytes"
	"context"
	"fmt"
	"net/http"
	"time"

	"github.com/golang/protobuf/proto"
	"github.com/golang/snappy"
	"github.com/rs/zerolog"

	"encore.dev/appruntime/config"
	"encore.dev/appruntime/metadata"
	"encore.dev/appruntime/metrics/prometheus/prompb"
	"encore.dev/appruntime/metrics/system"
	"encore.dev/metrics"
)

func New(svcs []string, cfg *config.PrometheusRemoteWriteProvider, meta *metadata.ContainerMetadata, rootLogger zerolog.Logger) *Exporter {
	// Precompute container metadata labels.
	return &Exporter{
		svcs: svcs,
		cfg:  cfg,
		containerMetadataLabels: []*prompb.Label{
			{Name: "service_id", Value: meta.ServiceID},
			{Name: "revision_id", Value: meta.RevisionID},
			{Name: "instance_id", Value: meta.InstanceID},
		},
		rootLogger: rootLogger,
	}
}

type Exporter struct {
	svcs                    []string
	cfg                     *config.PrometheusRemoteWriteProvider
	containerMetadataLabels []*prompb.Label
	rootLogger              zerolog.Logger
}

func (x *Exporter) Shutdown(_ context.Context) {}

func (x *Exporter) Export(ctx context.Context, collected []metrics.CollectedMetric) error {
	now := time.Now()
	data := x.getMetricData(now, collected)
	data = append(data, x.getSysMetrics(now)...)
	proto, err := proto.Marshal(&prompb.WriteRequest{Timeseries: data})
	if err != nil {
		return fmt.Errorf("unable to marshal metrics into Protobuf: %v", err)
	}

	encoded := snappy.Encode(nil, proto)
	body := bytes.NewReader(encoded)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, x.cfg.RemoteWriteURL, body)
	if err != nil {
		return fmt.Errorf("unable to create HTTP request: %v", err)
	}

	req.Header.Set("Content-Type", "application/x-protobuf")
	req.Header.Set("Content-Encoding", "snappy")
	req.Header.Set("User-Agent", "encore")
	req.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
	_, err = http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("unable to send metrics to Prometheus remote write destination: %v", err)
	}

	return nil
}

func (x *Exporter) getMetricData(now time.Time, collected []metrics.CollectedMetric) []*prompb.TimeSeries {
	data := make([]*prompb.TimeSeries, 0, len(collected))

	doAdd := func(val float64, metricName string, baseLabels []*prompb.Label, svcIdx uint16) {
		labels := make([]*prompb.Label, len(baseLabels)+2)
		copy(labels, baseLabels)
		labels[len(baseLabels)] = &prompb.Label{Name: "__name__", Value: metricName}
		labels[len(baseLabels)+1] = &prompb.Label{Name: "service", Value: x.svcs[svcIdx]}
		data = append(data, &prompb.TimeSeries{
			Labels: labels,
			Samples: []*prompb.Sample{
				{
					Value:     val,
					Timestamp: FromTime(now),
				},
			},
		})
	}

	for _, m := range collected {
		labels := make([]*prompb.Label, 0, len(x.containerMetadataLabels)+len(m.Labels))
		copy(labels, x.containerMetadataLabels)
		for _, label := range m.Labels {
			labels = append(labels, &prompb.Label{
				Name:  label.Key,
				Value: label.Value,
			})
		}

		svcNum := m.Info.SvcNum()
		switch vals := m.Val.(type) {
		case []float64:
			if svcNum > 0 {
				if m.Valid[0].Load() {
					doAdd(vals[0], m.Info.Name(), labels, svcNum-1)
				}
			} else {
				for i, val := range vals {
					if m.Valid[i].Load() {
						doAdd(val, m.Info.Name(), labels, uint16(i))
					}
				}
			}
		case []int64:
			if svcNum > 0 {
				if m.Valid[0].Load() {
					doAdd(float64(vals[0]), m.Info.Name(), labels, svcNum-1)
				}
			} else {
				for i, val := range vals {
					if m.Valid[i].Load() {
						doAdd(float64(val), m.Info.Name(), labels, uint16(i))
					}
				}
			}
		case []uint64:
			if svcNum > 0 {
				if m.Valid[0].Load() {
					doAdd(float64(vals[0]), m.Info.Name(), labels, svcNum-1)
				}
			} else {
				for i, val := range vals {
					if m.Valid[i].Load() {
						doAdd(float64(val), m.Info.Name(), labels, uint16(i))
					}
				}
			}
		case []time.Duration:
			if svcNum > 0 {
				if m.Valid[0].Load() {
					doAdd(float64(vals[0]/time.Second), m.Info.Name(), labels, svcNum-1)
				}
			} else {
				for i, val := range vals {
					if m.Valid[i].Load() {
						doAdd(float64(val/time.Second), m.Info.Name(), labels, uint16(i))
					}
				}
			}
		default:
			x.rootLogger.Error().Msgf("encore: internal error: unknown value type %T for metric %s",
				m.Val, m.Info.Name())
		}
	}

	return data
}

func (x *Exporter) getSysMetrics(now time.Time) []*prompb.TimeSeries {
	addMetricNameLabel := func(metricName string) []*prompb.Label {
		labels := make([]*prompb.Label, 0, len(x.containerMetadataLabels)+1)
		copy(labels, x.containerMetadataLabels)
		labels[len(x.containerMetadataLabels)] = &prompb.Label{
			Name:  "__name__",
			Value: metricName,
		}
		return labels
	}

	sysMetrics := system.ReadSysMetrics(x.rootLogger)
	return []*prompb.TimeSeries{
		{
			Labels: addMetricNameLabel(system.MetricNameHeapObjectsBytes),
			Samples: []*prompb.Sample{{
				Value:     float64(sysMetrics[system.MetricNameHeapObjectsBytes]),
				Timestamp: FromTime(now),
			}},
		},
		{
			Labels: addMetricNameLabel(system.MetricNameGoroutines),
			Samples: []*prompb.Sample{{
				Value:     float64(sysMetrics[system.MetricNameGoroutines]),
				Timestamp: FromTime(now),
			}},
		},
	}
}

// FromTime returns a new millisecond timestamp from a time.
func FromTime(t time.Time) int64 {
	return t.Unix()*1000 + int64(t.Nanosecond())/int64(time.Millisecond)
}

File: runtimes/core/src/metrics/exporter/prometheus.rs
use crate::encore::runtime::v1 as pb;
use crate::metadata::ContainerMetaClient;
use crate::metrics::exporter::Exporter;
use crate::metrics::{CollectedMetric, MetricValue};
use crate::secrets;
use anyhow::Context;
use prost::Message;
use std::sync::Arc;
use std::time::SystemTime;
use tokio::sync::OnceCell;
use url::Url;

#[derive(Debug)]
pub struct Prometheus {
    client: reqwest::Client,
    remote_write_url: Url,
    container_meta_client: ContainerMetaClient,
    container_labels: OnceCell<Arc<Vec<prompb::Label>>>,
}

impl Prometheus {
    pub fn new(
        provider_cfg: &pb::metrics_provider::PrometheusRemoteWrite,
        secrets: &secrets::Manager,
        container_meta_client: ContainerMetaClient,
    ) -> anyhow::Result<Self> {
        let remote_write_url = match &provider_cfg.remote_write_url {
            Some(data) => {
                let secret = secrets.load(data.clone());
                let url_bytes = secret
                    .get()
                    .context("failed to resolve Prometheus Remote Write Url")?;
                let s = std::str::from_utf8(url_bytes)
                    .context("Prometheus Remote Write Url is not valid UTF-8")?;
                Url::parse(s).context("Prometheus Remote Write Url not valid")?
            }
            None => {
                return Err(anyhow::anyhow!("Prometheus Remote Write Url not provided"));
            }
        };

        Ok(Self {
            client: reqwest::Client::new(),
            remote_write_url,
            container_meta_client,
            container_labels: OnceCell::new(),
        })
    }

    async fn export_metrics(&self, metrics: Vec<CollectedMetric>) -> anyhow::Result<()> {
        if metrics.is_empty() {
            return Ok(());
        }

        log::trace!(
            "Exporting {} metrics to Prometheus remote write host {}",
            metrics.len(),
            self.remote_write_url.host_str().unwrap_or_default()
        );

        let time_series = self.get_metric_data(metrics).await;

        // Create WriteRequest with the time series
        let write_request = prompb::WriteRequest {
            timeseries: time_series,
            metadata: vec![],
        };

        // Serialize to protobuf
        let mut proto_buf = Vec::new();
        write_request
            .encode(&mut proto_buf)
            .context("marshal metrics into Protobuf")?;

        // Compress with Snappy
        let mut encoder = snap::raw::Encoder::new();
        let encoded = encoder
            .compress_vec(&proto_buf)
            .context("compress metrics with Snappy")?;

        // Send HTTP request
        let response = self
            .client
            .post(self.remote_write_url.clone())
            .header("Content-Type", "application/x-protobuf")
            .header("Content-Encoding", "snappy")
            .header("User-Agent", "encore")
            .header("X-Prometheus-Remote-Write-Version", "0.1.0")
            .body(encoded)
            .send()
            .await
            .context("send metrics to Prometheus remote write destination")?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response
                .text()
                .await
                .unwrap_or_else(|_| "<unable to read response body>".to_string());
            anyhow::bail!(
                "Prometheus remote write returned non-success status {}: {}",
                status,
                body
            );
        }

        Ok(())
    }

    async fn container_labels(&self) -> Arc<Vec<prompb::Label>> {
        self.container_labels
            .get_or_try_init(|| async {
                let container_metadata = self.container_meta_client.collect().await?;
                anyhow::Ok(Arc::new(
                    container_metadata
                        .labels()
                        .into_iter()
                        .map(|(name, value)| prompb::Label { name, value })
                        .collect(),
                ))
            })
            .await
            .map(Arc::clone)
            .unwrap_or_else(|e| {
                log::warn!("failed fetching container metadata: {e}, using fallback");
                Arc::new(
                    self.container_meta_client
                        .fallback()
                        .labels()
                        .into_iter()
                        .map(|(name, value)| prompb::Label { name, value })
                        .collect(),
                )
            })
    }

    async fn get_metric_data(&self, collected: Vec<CollectedMetric>) -> Vec<prompb::TimeSeries> {
        let now = SystemTime::now();
        let timestamp = from_time(now);
        let mut data: Vec<prompb::TimeSeries> = Vec::with_capacity(collected.len());

        let container_labels = self.container_labels().await;
        let container_labels_len = container_labels.len();

        for metric in collected {
            let metric_name = metric.key.name().to_string();

            // Build labels: container metadata + metric labels + __name__
            let mut labels: Vec<prompb::Label> =
                Vec::with_capacity(container_labels_len + metric.key.labels().len() + 1);

            // Add container metadata labels
            labels.extend(container_labels.iter().cloned());

            // Add metric-specific labels
            for label in metric.key.labels() {
                labels.push(prompb::Label {
                    name: label.key().to_string(),
                    value: label.value().to_string(),
                });
            }

            // Add __name__ label for the metric name
            labels.push(prompb::Label {
                name: "__name__".to_string(),
                value: metric_name,
            });

            // Convert metric value to float64
            let value = match metric.value {
                MetricValue::CounterU64(val) => val as f64,
                MetricValue::CounterI64(val) => val as f64,
                MetricValue::GaugeF64(val) => val,
                MetricValue::GaugeU64(val) => val as f64,
                MetricValue::GaugeI64(val) => val as f64,
            };

            data.push(prompb::TimeSeries {
                labels,
                samples: vec![prompb::Sample { value, timestamp }],
                exemplars: vec![],
                histograms: vec![],
            });
        }

        data
    }
}

#[async_trait::async_trait]
impl Exporter for Prometheus {
    async fn export(&self, metrics: Vec<CollectedMetric>) {
        if let Err(err) = self.export_metrics(metrics).await {
            log::error!("Failed to export metrics to Prometheus: {}", err);
        }
    }
}

/// Convert SystemTime to Prometheus timestamp (milliseconds since Unix epoch)
fn from_time(t: SystemTime) -> i64 {
    match t.duration_since(SystemTime::UNIX_EPOCH) {
        Ok(duration) => {
            let secs = duration.as_secs() as i64;
            let nanos = duration.subsec_nanos() as i64;
            secs * 1000 + nanos / 1_000_000
        }
        Err(_) => 0, // If time is before Unix epoch, return 0
    }
}

#[allow(dead_code)]
mod prompb {
    include!(concat!(env!("OUT_DIR"), "/prometheus.rs"));
}
